src/backend/commands/matview.c

```diff
@@ -1763,14 +1897,160 @@ rewrite_query_for_counting_and_aggregation(Query *query, ParseState *pstate)
+				ParseState *pstate;
+				RangeTblEntry *rte;
+				RangeTblRef *rtr;
+				Alias *alias;
+
+				TargetEntry *tle_count;
+				FuncCall *fn;
+				Node *fn_node;
+				Expr *opexpr;
+
+				SubLink *sublink = (SubLink *)node;
+				/* return NULL if not has exist clause */
+				if (sublink->subLinkType != EXISTS_SUBLINK)
+					ereport(ERROR, (errmsg("subquery is not supported with IVM, except for EXISTS clause")));
+
+				subselect = (Query *)sublink->subselect;
+				/* return NULL if it is CTE */
+				if (subselect->cteList)
+					ereport(ERROR, (errmsg("CTE is not supported with IVM")));
+
+				pstate = make_parsestate(NULL);
+				pstate->p_expr_kind = EXPR_KIND_SELECT_TARGET;
+
+				/*
+				 * convert EXISTS subquery into LATERAL subquery in FROM clause.
+				 */
+
+
+				snprintf(aliasName, sizeof(aliasName), "__ivm_exists_subquery_%d__", *count);
+				snprintf(columnName, sizeof(columnName), "__ivm_exists_count_%d__", *count);
+
+				/* add COUNT(*) for counting exists condition */
+				fn = makeFuncCall(list_make1(makeString("count")), NIL, -1);
+				fn->agg_star = true;
+				fn_node = ParseFuncOrColumn(pstate, fn->funcname, NIL, NULL, fn, false, -1);
+				tle_count = makeTargetEntry((Expr *) fn_node,
+											list_length(subselect->targetList) + 1,
+											columnName,
+											false);
+				/* add __ivm_exists_count__ column */
+				subselect->targetList = list_concat(subselect->targetList, list_make1(tle_count));
+				subselect->hasAggs = true;
+
+				/* add subquery in from clause */
+				alias = makeAlias(aliasName, NIL);
+				/* it means that LATERAL is enable if fourth argument is true */
+				rte = addRangeTableEntryForSubquery(pstate,subselect,alias,true,true);
+
+				query->rtable = lappend(query->rtable, rte);
+				pstate->p_rtable = query->rtable; 
+
+				rtr = makeNode(RangeTblRef);
+				/* assume new rte is at end */
+				rtr->rtindex = list_length(query->rtable);
+
+				((FromExpr *)query->jointree)->fromlist = lappend(((FromExpr *)query->jointree)->fromlist, rtr);
+
+
+				fn = makeFuncCall(list_make1(makeString("count")), NIL, -1);
+				fn->agg_star = true;
+
+				fn_node = ParseFuncOrColumn(pstate, fn->funcname, NIL, NULL, fn, false, -1);
+
+				/*
+				 * it means using int84gt( '>' operator). it will be replaced to make_op().
+				 */
+				opexpr = make_opclause(419, BOOLOID, false,
+								(Expr *)fn_node,
+								(Expr *)makeConst(INT4OID, -1, InvalidOid, sizeof(int32), Int32GetDatum(0), false, true),
+								InvalidOid, InvalidOid);
+				fix_opfuncids((Node *) opexpr);
+				query->hasSubLinks = false;
+
+				subselect->havingQual = (Node *)opexpr;
+				(*count)++;
+				break;
+			}
+		default:
+			break;
+	}
+	return query;
+}
+
+
+Query *
+rewrite_query_for_exists_subquery(Query *query, Node *node)
+{
+	int count = 0;	
+	if (query->hasAggs)
+		elog(ERROR, "aggregate function and EXISTS condition are not supported at the same time");
+
+	return rewrite_exists_subquery_walker(query, node, &count);
```

Comment:
I think rewrite_query_for_exists_subquery can require only one argument because we can call the walker function like:
```
 return rewrite_exists_subquery_walker(query, (Node*) query, &count);
```

Reply:
You are right, I will change it to one argument.